File: apps/web/src/lib/bridge/commands/usage.ts
/** Bridge commands — the provider usage meter (issue #121). Thin wrappers over the
 *  `enable_usage_meter` / `disable_usage_meter` / `get_usage` / `refresh_usage` /
 *  `get_usage_cost` Tauri commands: a read-only telemetry poller that lives in the
 *  Rust core (it owns the Keychain + HTTP + credential-file seams). Outside the
 *  Tauri webview (browser preview / stories) they resolve quiet fallbacks so the
 *  sidebar widget degrades to its dormant "Enable usage meter" state instead of
 *  rejecting. */
import { tauriInvoke } from '../internal';
import type { UsageCost, UsageMeter, UsageStatus } from '../types';

/** The provider vocabulary the Rust registry pins (`usage::registry::PROVIDERS`) —
 *  used only to synthesize the browser-preview fallback meters below. */
const PREVIEW_PROVIDERS = ['claude', 'codex'] as const;

/** A synthesized meter with every provider in `status` — the browser-preview
 *  fallback for the read commands, mirroring the Rust `disabled_meter` /
 *  cold-registry shapes so the widget renders a stable layout without Tauri. */
function previewMeter(status: UsageStatus): UsageMeter {
  return {
    providers: PREVIEW_PROVIDERS.map((provider) => ({
      provider,
      status,
      windows: [],
      stale: false,
    })),
  };
}

/** The last-good snapshot (the widget's fetch-on-mount source of truth; no fetch
 *  Rust-side). Returns an all-`disabled` meter outside Tauri so the widget shows
 *  its opt-in "Enable usage meter" state in the browser preview. */
export async function getUsage(): Promise<UsageMeter> {
  return tauriInvoke<UsageMeter>('get_usage', {}, previewMeter('disabled'));
}

/** Opt in (spec decision 5): flips the persisted flag, performs the FIRST
 *  credential read (so the macOS Keychain prompt is a consequence of the click),
 *  arms the poll loop, and returns the initial snapshot. Outside Tauri it resolves
 *  an all-`notConnected` meter so the preview leaves the "Enable" state. */
export async function enableUsageMeter(): Promise<UsageMeter> {
  return tauriInvoke<UsageMeter>('enable_usage_meter', {}, previewMeter('notConnected'));
}

/** Opt out: flip the flag off; the Rust poll loop parks on the enable-kick. */
export async function disableUsageMeter(): Promise<void> {
  await tauriInvoke<null>('disable_usage_meter', {}, null);
}

/** Kick a fresh poll (the widget's `window` focus listener). Single-flight-guarded
 *  and internally no-ops unless the snapshot is ≥ 10 min stale, so a focus-storm
 *  can't hammer the endpoints. */
export async function refreshUsage(): Promise<void> {
  await tauriInvoke<null>('refresh_usage', {}, null);
}

/** The on-demand LOCAL cost estimate for a provider (spec §3.8) — invoked lazily
 *  when the detail popover opens, never on the 10-min poll. Always labeled
 *  approximate. Outside Tauri it resolves a null-cost estimate. */
export async function getUsageCost(provider: string): Promise<UsageCost> {
  return tauriInvoke<UsageCost>(
    'get_usage_cost',
    { provider },
    { provider, approximate: true, computedAt: new Date().toISOString() },
  );
}

File: apps/desktop/src-tauri/src/commands/usage.rs
//! The provider usage-meter command surface (issue #121) — thin async handlers over
//! [`crate::usage::UsageRegistry`] (managed state) + the `SettingsStore` opt-in flag.
//!
//! ALL async + `spawn_blocking` (a sync `#[tauri::command]` runs on the WKWebView
//! main thread and can freeze the UI — the known trap,
//! `reference_tauri_command_threading`; every terminal/trust command follows this).
//! State is re-acquired via `app.try_state()` inside each spawned block.
//!
//! USER-ONLY (spec §2): these commands are invokable only from the webview; no
//! agent/sidecar path reaches the meter.

use tauri::{AppHandle, Emitter, Manager};

use crate::settings::{SettingsPatch, SettingsStore};
use crate::usage::contract::{UsageCost, UsageMeter};
use crate::usage::{arm, kick, prime_credentials, UsageRegistry, REFRESH_MIN_AGE, USAGE_EVENT};

/// Opt in (spec decision 5): (a) flip the persisted `usage_meter_enabled` flag,
/// (b) perform the FIRST credential read so the macOS Keychain prompt fires as a
/// consequence of THIS click, (c) arm the recurring poll loop. Returns the current
/// snapshot (the first real windows arrive shortly over `nc:usage`). If the user
/// denies the Keychain prompt, the provider resolves to `Unauthorized`/`NotConnected`
/// — never a crash.
///
/// Also PUSHES that same snapshot over `nc:usage` (issue #305): the return value
/// only reaches the caller (e.g. the Settings toggle), but other mounted surfaces
/// (the sidebar widget) only ever render off the push — without this, an enable
/// triggered from Settings never reached the sidebar until the next 10-min poll.
#[tauri::command]
pub async fn enable_usage_meter(app: AppHandle) -> Result<UsageMeter, String> {
    tauri::async_runtime::spawn_blocking(move || {
        set_enabled(&app, true)?;
        // The credential read that fires the Keychain prompt (on the blocking pool —
        // a prompt may block). Result discarded; the loop does the real fetch.
        prime_credentials();
        // Arm (or wake) the loop; it polls immediately now that the flag is on.
        arm(&app);
        let reg = app
            .try_state::<UsageRegistry>()
            .ok_or("usage registry unavailable")?;
        let snapshot = reg.snapshot();
        let _ = app.emit(USAGE_EVENT, snapshot.clone());
        Ok(snapshot)
    })
    .await
    .map_err(|e| format!("enable usage meter failed to run: {e}"))?
}

/// Opt out: flip the flag off + wake the loop so it re-checks and parks (spending
/// zero network while disabled).
///
/// Also PUSHES a fresh `disabled_meter()` snapshot over `nc:usage` (issue #305): the
/// poll loop only parks silently on its own kick (`poller::run_loop`) — it never
/// emits on the disable path — so without this push, a surface that renders purely
/// off the `nc:usage` snapshot (the sidebar widget) stayed showing live data until
/// the next relaunch, even though the meter had just been turned off.
#[tauri::command]
pub async fn disable_usage_meter(app: AppHandle) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        set_enabled(&app, false)?;
        kick(&app);
        let _ = app.emit(USAGE_EVENT, UsageRegistry::disabled_meter());
        Ok(())
    })
    .await
    .map_err(|e| format!("disable usage meter failed to run: {e}"))?
}

/// The last-good snapshot (the fetch-on-mount source of truth; no fetch here).
/// Returns a `Disabled`-status meter when opt-in-off, so the widget renders its
/// "Enable usage meter" state.
#[tauri::command]
pub async fn get_usage(app: AppHandle) -> Result<UsageMeter, String> {
    tauri::async_runtime::spawn_blocking(move || {
        if !is_enabled(&app) {
            return Ok(UsageRegistry::disabled_meter());
        }
        let reg = app
            .try_state::<UsageRegistry>()
            .ok_or("usage registry unavailable")?;
        Ok(reg.snapshot())
    })
    .await
    .map_err(|e| format!("get usage failed to run: {e}"))?
}

/// Kick a fresh poll (the web `window` focus listener). Single-flight-guarded, and
/// internally no-ops unless the snapshot is ≥ 10 min old (spec §3.3 staleness guard)
/// so a focus-storm can't hammer the endpoints.
#[tauri::command]
pub async fn refresh_usage(app: AppHandle) -> Result<(), String> {
    tauri::async_runtime::spawn_blocking(move || {
        if let Some(reg) = app.try_state::<UsageRegistry>() {
            if is_enabled(&app) && reg.stale_enough(REFRESH_MIN_AGE) {
                kick(&app);
            }
        }
        Ok(())
    })
    .await
    .map_err(|e| format!("refresh usage failed to run: {e}"))?
}

/// The on-demand local cost ESTIMATE for a provider (spec §3.8) — invoked when the
/// detail popover opens. `spawn_blocking` (a whole-tree JSONL read). Always labeled
/// approximate in the result.
#[tauri::command]
pub async fn get_usage_cost(app: AppHandle, provider: String) -> Result<UsageCost, String> {
    tauri::async_runtime::spawn_blocking(move || {
        let reg = app
            .try_state::<UsageRegistry>()
            .ok_or("usage registry unavailable")?;
        Ok(reg.cost(&provider))
    })
    .await
    .map_err(|e| format!("get usage cost failed to run: {e}"))?
}

/// Flip the persisted `usage_meter_enabled` flag via the settings store.
fn set_enabled(app: &AppHandle, enabled: bool) -> Result<(), String> {
    let store = app
        .try_state::<SettingsStore>()
        .ok_or("settings store unavailable")?;
    store.update(SettingsPatch {
        usage_meter_enabled: Some(enabled),
        ..Default::default()
    })?;
    Ok(())
}

/// Whether the meter is opt-in-on. `false` when the settings store isn't managed.
fn is_enabled(app: &AppHandle) -> bool {
    app.try_state::<SettingsStore>()
        .map(|s| s.with_settings(|s| s.usage_meter_enabled))
        .unwrap_or(false)
}
